prog/dftb+/lib_io/ipisocket.F90

```diff
@@ -1,373 +1,373 @@
+      call error(msg)
+    end if
+
+    ! read actual coordinates
+    call readbuffer(this%socket, commsBuffer1)
+    coord = reshape(commsBuffer1, [3, natom])
+    call this%logger%write('ipisocket%receive: read from socket: atomic positions', 3)
+    call this%logger%write(coord, 5, '(f12.6)')
+    call this%logger%write('ipisocket%receive: Done', 1)
+
+  end subroutine receive
+
+
+  !> Send data to an external program via a socket
+  !!
+  !! All data in atomic units, and currently assumes the number
+  !! of atoms is the same as passed at construction/initialisation.
+  !!
+  subroutine send(this, energy, forces, stress)
+
+
+    !> Instance
+    class(IpiSocketComm), intent(inout) :: this
+
+
+    !> Total energy
+    real(dp), intent(in) :: energy
+
+
+    !> Total forces
+    real(dp), intent(in) :: forces(:,:)
+
+
+    !> Cell stresses
+    real(dp), intent(in) :: stress(3, 3)
+
+    character(len=IPI_MSGLEN) :: header, buffer
+    character(lc) :: msg
+    integer :: nAtom
+
+    @:ASSERT(this%tInit)
+    @:ASSERT(size(forces, dim=1) == 3)
+
+    nAtom = size(forces, dim=2)
+    if (nAtom /= this%nAtom) then
+      write(msg, '(1X,A,2I4)') 'Mismatch in number of atoms in socketSend', nAtom,this%nAtom
+      call error(msg)
+    end if
+
+    call this%logger%write('ipisocket%send: Sending data to socket... ', 1)
+
+    ! wait for anything other than a 'STATUS' state from the interface,
+    ! returning state 'HAVEDATA' in the meanwhile
+    listen: do while (.true.)
+      call readbuffer(this%socket, header)
+      call this%logger%write('ipisocket%send: read from socket: ' // trim(header), 3)
+
+      if (trim(header)/='STATUS') then
+        exit listen
+      end if
+
+      ! announce that we have available data
+      buffer = 'HAVEDATA'
+      call writebuffer(this%socket, buffer)
+      call this%logger%write('ipisocket%send: write to socket: HAVEDATA', 3)
+    end do listen
+
+    ! expecting to send force data
+    if (trim(header) /= 'GETFORCE') then
+      call error("ipisocket%send: Error in socket communication! (expected 'GETFORCE',&
+          & received '" // trim(header) // "'")
+    end if
+
+    buffer = 'FORCEREADY'
+    call writebuffer(this%socket, buffer)
+    call this%logger%write('ipisocket%send: write to socket: FORCEREADY', 3)
+
+    ! transmit total energy
+    call writebuffer(this%socket, energy)
+    call this%logger%write('ipisocket%send: write to socket: energy', 3)
+    write(msg, *) energy
+    call this%logger%write(msg, 4)
+
+    ! transmit number of atoms we have
+    call writebuffer(this%socket, natom)
+    call this%logger%write('ipisocket%send: write to socket: natom', 3)
+
+    ! transmit forces
+    call writebuffer(this%socket, reshape(forces, [3 * natom]))
+    call this%logger%write('ipisocket%send: write to socket: forces', 3)
+    call this%logger%write(forces, 5, '(f12.6)')
+
+    ! transmit stress
+    call writebuffer(this%socket, reshape(stress, [9]))

```

Comment:
To be consistent with the transpose of the returned lattice vectors, should this not also be transposed?

Reply:
I have indeed checked the stress tensor, and it was okay (without transposing): in i-pi I printed out "stress_md" that has the format [xx, yy, zz, xy, xz, yz], and it was consistent with the "Total stress tensor" printed out in detailed.out by dftb+.
To comfirm, the "Total stress tensor" from dftb+ follows the format
xx xy xz
yx yy yz
zx zy zz?

Reply:
The tensor is symmetric, so is probably hiding this. Its probably a good idea to either transpose (even though it is 'harmless'), or add a comment in line 345 that this is symmetric, so does not need transposition.

Looking at the (i-PI driver example)[https://github.com/i-pi/i-pi/blob/fae06f697c30436a38591320acffaf8213e85505/drivers/f90/driver.f90#L679] it looks like this _might_ be expecting the virial, not the stress tensor. Do you know anything about this?

Reply:
The stress tensor is indeed symmetric so it doesn't matter. I'll add a comment to note this.

I think the "virial" in i-pi refers to the fact that it's only the virial component of the stress tensor (excl the kinetic part). If there's an opposite sign because of the different defination of the internal/external pressure, this should show up when I printed out the "stress_md" using i-pi.